README.md
# pynbs

[![GitHub Actions](https://github.com/OpenNBS/pynbs/workflows/CI/badge.svg)](https://github.com/OpenNBS/pynbs/actions)
[![PyPI](https://img.shields.io/pypi/v/pynbs.svg)](https://pypi.org/project/pynbs/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/pynbs.svg)](https://pypi.org/project/pynbs/)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black)

> A simple python library to read and write [.nbs files](https://opennbs.org/nbs)
> from [Open Note Block Studio](https://opennbs.org/).

`pynbs` makes it possible to easily iterate over Note Block Studio songs.

```python
import pynbs

for tick, chord in pynbs.read('demo_song.nbs'):
    print(tick, [note.key for note in chord])
```

You can also use `pynbs` to generate new songs programmatically.

```python
import pynbs

new_file = pynbs.new_file(song_name='Hello world')
new_file.notes.extend([
    pynbs.Note(tick=i, layer=0, instrument=0, key=i + 35) for i in range(10)
])

new_file.save('new_file.nbs')
```

## Installation

The package can be installed with `pip`.

```bash
$ pip install pynbs
```

The latest release follows the latest version of the NBS file format
[specification](https://opennbs.org/nbs)
(version 5). However, it also allows you to load and save files in any of
the older versions.

## Basic usage

### Reading files

You can use the `read()` function to read and parse a specific NBS file.

```python
demo_song = pynbs.read('demo_song.nbs')
```

The `read()` function returns a `pynbs` file object. These objects have several
attributes that mirror the binary structure of NBS files.

#### Header

The first attribute is `header`, the file header. It contains information about
the file.

```python
header = demo_song.header
```

Attribute                   | Type    | Details
:---------------------------|:--------|:------------------------------------------------
`header.version`            | `int`   | The NBS version this file was saved on.
`header.default_instruments`| `int`   | The amount of instruments from vanilla Minecraft in the song.
`header.song_length`        | `int`   | The length of the song, measured in ticks.
`header.song_layers`        | `int`   | The ID of the last layer with at least one note block in it.
`header.song_name`          | `str`   | The name of the song.
`header.song_author`        | `str`   | The author of the song.
`header.original_author`    | `str`   | The original song author of the song.
`header.description`        | `str`   | The description of the song.
`header.tempo`              | `float` | The tempo of the song.
`header.auto_save`          | `bool`  | Whether auto-saving has been enabled.
`header.auto_save_duration` | `int`   | The amount of minutes between each auto-save.
`header.time_signature`     | `int`   | The time signature of the song.
`header.minutes_spent`      | `int`   | The amount of minutes spent on the project.
`header.left_clicks`        | `int`   | The amount of times the user has left-clicked.
`header.right_clicks`       | `int`   | The amount of times the user has right-clicked.
`header.blocks_added`       | `int`   | The amount of times the user has added a block.
`header.blocks_removed`     | `int`   | The amount of times the user has removed a block.
`header.song_origin`        | `str`   | The file name of the original MIDI or schematic.
`header.loop`               | `bool`  | Whether the song should loop back to the start after ending.
`header.max_loop_count`     | `int`   | The amount of times to loop. 0 = infinite.
`header.loop_start`         | `int`   | The tick the song will loop back to at the end of playback.

> For more information about all these fields, check out the [official specification](https://hielkeminecraft.github.io/OpenNoteBlockStudio/nbs).

#### Notes

The `notes` attribute holds a list of all the notes of the song in order.

```python
first_note = demo_song.notes[0]
```

Attribute         | Type  | Details
:---------------- |:------|:------------------------------------------------
`note.tick`       | `int` | The tick at which the note plays.
`note.layer`      | `int` | The ID of the layer in which the note is placed.
`note.instrument` | `int` | The ID of the instrument.
`note.key`        | `int` | The key of the note. (between 0 and 87)
`note.velocity`   | `int` | The velocity of the note. (between 0 and 100)
`note.panning`    | `int` | The stereo panning of the note. (between -100 and 100)
`note.pitch`      | `int` | The detune of the note, in cents. (between -1200 and 1200)

#### Layers

The `layers` attribute holds a list of all the layers of the song in order.

```python
first_layer = demo_song.layers[0]
```

Attribute         | Type  | Details
:-----------------|:------|:------------------------
`layer.id`        | `int` | The ID of the layer.
`layer.name`      | `str` | The name of the layer.
`layer.lock`      | `bool`| Whether the layer is locked.
`layer.volume`    | `int` | The volume of the layer.
`layer.panning`   | `int` | The stereo panning of the layer.

#### Instruments

The `instruments` attribute holds a list of all the custom instruments of the
song in order.

```python
first_custom_instrument = demo_song.instruments[0]
```

Attribute              | Type   | Details
:----------------------|:-------|:----------------------------------------------------------
`instrument.id`        | `int`  | The ID of the instrument.
`instrument.name`      | `str`  | The name of the instrument.
`instrument.file`      | `str`  | The name of the sound file of the instrument.
`instrument.pitch`     | `int`  | The pitch of the instrument. (between 0 and 87)
`instrument.press_key` | `bool` | Whether the piano should automatically press keys with the instrument when the marker passes them.

### Iterating over songs

Iterating over a `pynbs` file object yields consecutively all the chords of the song with
the associated tick.

```python
for tick, chord in demo_song:
    ...
```

`chord` is a list of all the notes that play during the tick `tick`.

### Creating new files

You can create new files using the `new_file()` function. The function lets
you specify header attributes with keyword arguments.

```python
new_file = pynbs.new_file(song_name='Hello world')
```

The function returns a new `pynbs` file object that you can now edit
programmatically.

### Saving files

You can use the `save()` method to encode and write the file to a specified
location.

```python
new_file.save('new_file.nbs')
```

By default, the file will be saved in the latest NBS version available.
To save the file in an older version, you can use the `version` parameter:

```python
# This will save the song in the classic format.
new_file.save('new_file.nbs', version=0)
```

(Keep in mind some of the song properties may be lost when saving in older versions.)

### Upgrading old files

While `pynbs` is up-to-date with the latest version of the Open Note Block Studio
specification, all previous versions — including the original file format — are still
supported by the `read()` function, making it possible to bulk upgrade songs to the
most recent version:

```python
import glob
import pynbs

for old_file in glob.glob('*.nbs'):
    pynbs.read(old_file).save(old_file)
```

## Contributing

Contributions are welcome. Make sure to first open an issue discussing the problem or the new feature before creating a pull request. The project uses [`poetry`](https://python-poetry.org/).

```bash
$ poetry install
```

You can run the tests with `poetry run pytest`.

```bash
$ poetry run pytest
```

The code follows the [`black`](https://github.com/psf/black) code style. Import statements are sorted with [`isort`](https://pycqa.github.io/isort/).

```bash
$ poetry run isort pynbs tests
$ poetry run black pynbs tests
$ poetry run black --check pynbs tests
```

---

License - [MIT](https://github.com/OpenNBS/pynbs/blob/master/LICENSE)


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.1.0-beta.0 (2022-04-10)


## v1.1.0 (2022-04-10)
### Feature
* Publish 1.0 for real ([`fe8fcac`](https://github.com/OpenNBS/pynbs/commit/fe8fcac4397593d65a9109300e20784863fe32e5))

## v1.0.0-beta.4 (2022-04-10)
### Fix
* Remove note about python 2 compatibility ([`677121a`](https://github.com/OpenNBS/pynbs/commit/677121a6ab086976261dc1031bc83d6175564fca))

## v1.0.0-beta.3 (2022-04-10)
### Fix
* Remove py.typed for now ([`440c96b`](https://github.com/OpenNBS/pynbs/commit/440c96b6152d73b9c03f809ccd1615b723d9345b))
* Rename buffer to fileobj ([`0f9a379`](https://github.com/OpenNBS/pynbs/commit/0f9a37913acbdc9b7471966d330312cb90d9664c))
* Leverage Header constructor directly ([`5483f48`](https://github.com/OpenNBS/pynbs/commit/5483f48085aa991dd697731a036cbbf0a2982bfd))
* Inheriting from object not needed in python3 ([`b7dd7a2`](https://github.com/OpenNBS/pynbs/commit/b7dd7a27b63a89acbc004c38ea5af67a2756962c))
* Convert namedtuples to dataclasses ([`bad6dc9`](https://github.com/OpenNBS/pynbs/commit/bad6dc90c8f1ca951575d97f9487ed59d6794431))
* Move code out of `__init__` ([`6dd5f98`](https://github.com/OpenNBS/pynbs/commit/6dd5f98362af823d9fc74776edd243b50aea32c8))
* Add badges and constributing section to README ([`13d823b`](https://github.com/OpenNBS/pynbs/commit/13d823b66d4bb1175b00fbbd22383173e3ba9eb6))
* Put test.nbs into test resources ([`38316dc`](https://github.com/OpenNBS/pynbs/commit/38316dc4a4eb76d8edd394b780a2eb1dc60b6d35))

## v1.0.0-beta.2 (2022-04-10)
### Fix
* Run tests in CI ([`b8a3eab`](https://github.com/OpenNBS/pynbs/commit/b8a3eabc4891bfba496e199942c07392b153b549))
* Turn demo into tests ([`837aef2`](https://github.com/OpenNBS/pynbs/commit/837aef2053bc433b4c26edb1804151332f8ea244))
* Rename examples folder ([`009f4a9`](https://github.com/OpenNBS/pynbs/commit/009f4a9efb91c5b7c7a3e9c8c69b617a8d6f64cf))
* Update authors ([`a8ec277`](https://github.com/OpenNBS/pynbs/commit/a8ec277e5f5c8a3d01add12fe92c6babb3c2e566))

## v1.0.0-beta.1 (2022-04-10)
### Feature
* Start improving project structure ([`5949476`](https://github.com/OpenNBS/pynbs/commit/5949476596f8cc3ba22e366c40c0e54b570dea5b))

### Fix
* Integrate changes from main and target version 1.0 ([`07ab0f8`](https://github.com/OpenNBS/pynbs/commit/07ab0f887bbd2fd820118e4d4b6945ecbbeb3c80))

## v0.5.0-beta.0 (2022-04-07)
### Feature
* Start improving project structure ([`47147d2`](https://github.com/OpenNBS/pynbs/commit/47147d25079cac7084663b71fc285a5b11629173))


tests/conftest.py
from pathlib import Path

from pytest_insta import Fmt


class FmtNbs(Fmt[bytes]):
    extension = ".nbs"

    def load(self, path: Path) -> bytes:
        return path.read_bytes()

    def dump(self, path: Path, value: bytes):
        path.write_bytes(value)


tests/test_song.py
import os
from pathlib import Path

import pytest
from pytest_insta import SnapshotFixture

import pynbs


@pytest.mark.parametrize("filename", os.listdir("tests/resources"))
def test_notes(snapshot: SnapshotFixture, filename: str):
    f = pynbs.read(f"tests/resources/{filename}")
    data = "".join(f"{tick}: {[note.key for note in chord]}\n" for tick, chord in f)
    assert snapshot() == (
        f"{f.header.song_length = }\n"
        f"{f.header.description = }\n"
        f"{len(f.notes) = }\n"
        f"{len(f.layers) = }\n"
        f"{len(f.instruments) = }\n"
        f"{data}"
    )


def test_create(snapshot: SnapshotFixture, tmp_path: Path):
    f = pynbs.new_file(song_name="foo", song_author="bar")

    f.notes.extend(
        [
            pynbs.Note(tick=0, layer=0, instrument=0, key=45),
            pynbs.Note(tick=2, layer=0, instrument=0, key=45),
            pynbs.Note(tick=4, layer=0, instrument=0, key=45),
            pynbs.Note(tick=6, layer=0, instrument=0, key=45),
            pynbs.Note(tick=8, layer=0, instrument=0, key=45),
        ]
    )

    f.header.blocks_added = 9000

    f.save(tmp_path / "new.nbs")
    f.save(tmp_path / "old.nbs", version=0)

    new = (tmp_path / "new.nbs").read_bytes()
    old = (tmp_path / "old.nbs").read_bytes()
    assert new != old

    assert snapshot("new.nbs") == new
    assert snapshot("old.nbs") == old


pynbs/file.py
__all__ = [
    "read",
    "new_file",
    "Parser",
    "Writer",
    "File",
    "Header",
    "Note",
    "Layer",
    "Instrument",
]


from dataclasses import dataclass
from struct import Struct

CURRENT_NBS_VERSION = 5

BYTE = Struct("<B")
SHORT = Struct("<H")
SSHORT = Struct("<h")
INT = Struct("<I")


@dataclass
class Instrument:
    id: int
    name: str
    file: str
    pitch: int = 45
    press_key: bool = True


@dataclass
class Note:
    tick: int
    layer: int
    instrument: int
    key: int
    velocity: int = 100
    panning: int = 0
    pitch: int = 0


@dataclass
class Layer:
    id: int
    name: str = ""
    lock: bool = False
    volume: int = 100
    panning: int = 0


def read(filename):
    with open(filename, "rb") as fileobj:
        return Parser(fileobj).read_file()


def new_file(**header):
    return File(Header(**header), [], [Layer(0, "", False, 100, 0)], [])


@dataclass
class Header:
    version: int = CURRENT_NBS_VERSION
    default_instruments: int = 16
    song_length: int = 0
    song_layers: int = 0
    song_name: str = ""
    song_author: str = ""
    original_author: str = ""
    description: str = ""
    tempo: float = 10.0
    auto_save: bool = False
    auto_save_duration: int = 10
    time_signature: int = 4
    minutes_spent: int = 0
    left_clicks: int = 0
    right_clicks: int = 0
    blocks_added: int = 0
    blocks_removed: int = 0
    song_origin: str = ""
    loop: bool = False
    max_loop_count: int = 0
    loop_start: int = 0


class File:
    def __init__(self, header, notes, layers, instruments):
        self.header = header
        self.notes = notes
        self.layers = layers
        self.instruments = instruments

    def update_header(self, version):
        self.header.version = version
        if self.notes:
            self.header.song_length = self.notes[-1].tick
        self.header.song_layers = len(self.layers)

    def save(self, filename, version=CURRENT_NBS_VERSION):
        self.update_header(version)
        with open(filename, "wb") as fileobj:
            Writer(fileobj).encode_file(self, version)

    def __iter__(self):
        if not self.notes:
            return
        chord = []
        current_tick = self.notes[0].tick

        for note in sorted(self.notes, key=lambda n: n.tick):
            if note.tick == current_tick:
                chord.append(note)
            else:
                chord.sort(key=lambda n: n.layer)
                yield current_tick, chord
                current_tick, chord = note.tick, [note]
        yield current_tick, chord


class Parser:
    def __init__(self, fileobj):
        self.fileobj = fileobj

    def read_file(self):
        header = self.parse_header()
        version = header.version
        return File(
            header,
            list(self.parse_notes(version)),
            list(self.parse_layers(header.song_layers, version)),
            list(self.parse_instruments(version)),
        )

    def read_numeric(self, fmt):
        return fmt.unpack(self.fileobj.read(fmt.size))[0]

    def read_string(self):
        length = self.read_numeric(INT)
        return self.fileobj.read(length).decode(encoding="cp1252")

    def jump(self):
        value = -1
        while True:
            jump = self.read_numeric(SHORT)
            if not jump:
                break
            value += jump
            yield value

    def parse_header(self):
        song_length = self.read_numeric(SHORT)
        if song_length == 0:
            # A song length of 0 indicates the Open Note Block Studio format
            version = self.read_numeric(BYTE)
        else:
            version = 0

        return Header(
            version=version,
            default_instruments=self.read_numeric(BYTE) if version > 0 else 10,
            song_length=self.read_numeric(SHORT) if version >= 3 else song_length,
            song_layers=self.read_numeric(SHORT),
            song_name=self.read_string(),
            song_author=self.read_string(),
            original_author=self.read_string(),
            description=self.read_string(),
            tempo=self.read_numeric(SHORT) / 100.0,
            auto_save=self.read_numeric(BYTE) == 1,
            auto_save_duration=self.read_numeric(BYTE),
            time_signature=self.read_numeric(BYTE),
            minutes_spent=self.read_numeric(INT),
            left_clicks=self.read_numeric(INT),
            right_clicks=self.read_numeric(INT),
            blocks_added=self.read_numeric(INT),
            blocks_removed=self.read_numeric(INT),
            song_origin=self.read_string(),
            loop=self.read_numeric(BYTE) == 1 if version >= 4 else False,
            max_loop_count=self.read_numeric(BYTE) if version >= 4 else 0,
            loop_start=self.read_numeric(SHORT) if version >= 4 else 0,
        )

    def parse_notes(self, version):
        for current_tick in self.jump():
            for current_layer in self.jump():
                instrument = self.read_numeric(BYTE)
                key = self.read_numeric(BYTE)
                velocity = self.read_numeric(BYTE) if version >= 4 else 100
                panning = self.read_numeric(BYTE) - 100 if version >= 4 else 0
                pitch = self.read_numeric(SSHORT) if version >= 4 else 0
                yield Note(
                    current_tick,
                    current_layer,
                    instrument,
                    key,
                    velocity,
                    panning,
                    pitch,
                )

    def parse_layers(self, layers_count, version):
        for i in range(layers_count):
            name = self.read_string()
            lock = self.read_numeric(BYTE) == 1 if version >= 4 else False
            volume = self.read_numeric(BYTE)
            panning = self.read_numeric(BYTE) - 100 if version >= 2 else 0
            yield Layer(i, name, lock, volume, panning)

    def parse_instruments(self, version):
        for i in range(self.read_numeric(BYTE)):
            name = self.read_string()
            sound_file = self.read_string()
            pitch = self.read_numeric(BYTE)
            press_key = self.read_numeric(BYTE) == 1
            yield Instrument(i, name, sound_file, pitch, press_key)


class Writer:
    def __init__(self, fileobj):
        self.fileobj = fileobj

    def encode_file(self, nbs_file, version):
        self.write_header(nbs_file, version)
        self.write_notes(nbs_file, version)
        self.write_layers(nbs_file, version)
        self.write_instruments(nbs_file, version)

    def encode_numeric(self, fmt, value):
        self.fileobj.write(fmt.pack(value))

    def encode_string(self, value):
        self.encode_numeric(INT, len(value))
        self.fileobj.write(value.encode(encoding="cp1252"))

    def write_header(self, nbs_file, version):
        header = nbs_file.header

        if version > 0:
            self.encode_numeric(SHORT, 0)
            self.encode_numeric(BYTE, version)
            self.encode_numeric(BYTE, header.default_instruments)
        else:
            self.encode_numeric(SHORT, header.song_length)
        if version >= 3:
            self.encode_numeric(SHORT, header.song_length)
        self.encode_numeric(SHORT, header.song_layers)
        self.encode_string(header.song_name)
        self.encode_string(header.song_author)
        self.encode_string(header.original_author)
        self.encode_string(header.description)

        self.encode_numeric(SHORT, int(header.tempo * 100))
        self.encode_numeric(BYTE, int(header.auto_save))
        self.encode_numeric(BYTE, header.auto_save_duration)
        self.encode_numeric(BYTE, header.time_signature)

        self.encode_numeric(INT, header.minutes_spent)
        self.encode_numeric(INT, header.left_clicks)
        self.encode_numeric(INT, header.right_clicks)
        self.encode_numeric(INT, header.blocks_added)
        self.encode_numeric(INT, header.blocks_removed)
        self.encode_string(header.song_origin)

        if version >= 4:
            self.encode_numeric(BYTE, int(header.loop))
            self.encode_numeric(BYTE, header.max_loop_count)
            self.encode_numeric(SHORT, header.loop_start)

    def write_notes(self, nbs_file, version):
        current_tick = -1

        for tick, chord in nbs_file:
            self.encode_numeric(SHORT, tick - current_tick)
            current_tick = tick
            current_layer = -1

            for note in chord:
                self.encode_numeric(SHORT, note.layer - current_layer)
                current_layer = note.layer
                self.encode_numeric(BYTE, note.instrument)
                self.encode_numeric(BYTE, note.key)
                if version >= 4:
                    self.encode_numeric(BYTE, note.velocity)
                    self.encode_numeric(BYTE, note.panning + 100)
                    self.encode_numeric(SSHORT, note.pitch)

            self.encode_numeric(SHORT, 0)
        self.encode_numeric(SHORT, 0)

    def write_layers(self, nbs_file, version):
        for layer in nbs_file.layers:
            self.encode_string(layer.name)
            if version >= 4:
                self.encode_numeric(BYTE, int(layer.lock))
            self.encode_numeric(BYTE, layer.volume)
            if version >= 2:
                self.encode_numeric(BYTE, layer.panning + 100)

    def write_instruments(self, nbs_file, version):
        self.encode_numeric(BYTE, len(nbs_file.instruments))
        for instrument in nbs_file.instruments:
            self.encode_string(instrument.name)
            self.encode_string(instrument.file)
            self.encode_numeric(BYTE, instrument.pitch)
            self.encode_numeric(BYTE, int(instrument.press_key))


examples/demo.py
from __future__ import print_function

import pynbs


# Read file

for song_name in ['demo_song.nbs', 'old_demo_song.nbs']:
    demo_song = pynbs.read(song_name)

    print(demo_song.header.song_length)
    print(demo_song.header.description)

    print(demo_song.notes)
    print(demo_song.layers)
    print(demo_song.instruments)

    for tick, chord in demo_song:
        print(tick, [note.key for note in chord])


# Create new file

new_file = pynbs.new_file(
    song_name='foo',
    song_author='bar',
)


# Edit file

new_file.notes.extend([
    pynbs.Note(tick=0, layer=0, instrument=0, key=45),
    pynbs.Note(tick=2, layer=0, instrument=0, key=45),
    pynbs.Note(tick=4, layer=0, instrument=0, key=45),
    pynbs.Note(tick=6, layer=0, instrument=0, key=45),
    pynbs.Note(tick=8, layer=0, instrument=0, key=45),
])

new_file.header.blocks_added = 9000


# Save file

new_file.save('new_file.nbs')



# Save file in older version

new_file.save('old_new_file.nbs', version=0)


pynbs.py
from __future__ import print_function, division, absolute_import

try:
    from builtins import range
except ImportError:
    range = xrange

from struct import Struct
from collections import namedtuple


__all__ = ['read', 'new_file', 'Parser', 'Writer', 'File', 'Header',
           'Note', 'Layer', 'Instrument']

CURRENT_NBS_VERSION = 5

BYTE = Struct('<B')
SHORT = Struct('<H')
SSHORT = Struct('<h')
INT = Struct('<I')



Instrument = namedtuple('Instrument', ['id', 'name', 'file', 'pitch',
                                       'press_key'])


class Note(namedtuple('Note', ['tick', 'layer', 'instrument', 'key',
                      'velocity', 'panning', 'pitch'])):
    def __new__(cls, tick, layer, instrument, key, velocity=100,
               panning=0, pitch=0):
        return super().__new__(cls, tick, layer, instrument, key,
                               velocity, panning, pitch)

class Layer(namedtuple('Layer', ['id', 'name', 'lock', 'volume', 'panning'])):
    def __new__(cls, id, name="", lock=False, volume=100, panning=0):
        return super().__new__(cls, id, name, lock, volume, panning)


def read(filename):
    with open(filename, 'rb') as buff:
        return Parser(buff).read_file()


def new_file(**header):
    return File(Header(**header), [], [Layer(0, '', False, 100, 0)], [])


class Header(object):
    def __init__(self, **header):
        header_values = {
            'version':              header.get('version', CURRENT_NBS_VERSION),
            'default_instruments':  header.get('default_instruments', 16),
            'song_length':          header.get('song_length', 0),
            'song_layers':          header.get('song_layers', 0),
            'song_name':            header.get('song_name', ''),
            'song_author':          header.get('song_author', ''),
            'original_author':      header.get('original_author', ''),
            'description':          header.get('description', ''),

            'tempo':                header.get('tempo', 10.0),
            'auto_save':            header.get('auto_save', False),
            'auto_save_duration':   header.get('auto_save_duration', 10),
            'time_signature':       header.get('time_signature', 4),

            'minutes_spent':        header.get('minutes_spent', 0),
            'left_clicks':          header.get('left_clicks', 0),
            'right_clicks':         header.get('right_clicks', 0),
            'blocks_added':         header.get('blocks_added', 0),
            'blocks_removed':       header.get('blocks_removed', 0),
            'song_origin':          header.get('song_origin', ''),

            'loop':                 header.get('loop', False),
            'max_loop_count':       header.get('max_loop_count', 0),
            'loop_start':           header.get('loop_start', 0)
        }

        for key, value in header_values.items():
            setattr(self, key, value)


class File(object):
    def __init__(self, header, notes, layers, instruments):
        self.header = header
        self.notes = notes
        self.layers = layers
        self.instruments = instruments

    def update_header(self, version):
        self.header.version = version
        if self.notes:
            self.header.song_length = self.notes[-1].tick
        self.header.song_layers = len(self.layers)

    def save(self, filename, version=CURRENT_NBS_VERSION):
        self.update_header(version)
        with open(filename, 'wb') as buff:
            Writer(buff).encode_file(self, version)

    def __iter__(self):
        if not self.notes:
            return
        chord = []
        current_tick = self.notes[0].tick

        for note in sorted(self.notes, key=lambda n: n.tick):
            if note.tick == current_tick:
                chord.append(note)
            else:
                chord.sort(key=lambda n: n.layer)
                yield current_tick, chord
                current_tick, chord = note.tick, [note]
        yield current_tick, chord


class Parser(object):
    def __init__(self, buff):
        self.buffer = buff

    def read_file(self):
        values = self.parse_header()
        header = Header(**values)
        version = values['version']
        return File(header, list(self.parse_notes(version)),
                    list(self.parse_layers(header.song_layers, version)),
                    list(self.parse_instruments(version)))

    def read_numeric(self, fmt):
        return fmt.unpack(self.buffer.read(fmt.size))[0]

    def read_string(self):
        length = self.read_numeric(INT)
        return self.buffer.read(length).decode(encoding='cp1252')

    def jump(self):
        value = -1
        while True:
            jump = self.read_numeric(SHORT)
            if not jump:
                break
            value += jump
            yield value

    def parse_header(self):
        song_length = self.read_numeric(SHORT)
        if song_length == 0:
            # A song length of 0 indicates the Open Note Block Studio format
            version = self.read_numeric(BYTE)
        else:
            version = 0

        return {
            'version':             version,
            'default_instruments': self.read_numeric(BYTE)  if version > 0 else 10,
            'song_length':         self.read_numeric(SHORT) if version >= 3 else song_length,
            'song_layers':         self.read_numeric(SHORT),
            'song_name':           self.read_string(),
            'song_author':         self.read_string(),
            'original_author':     self.read_string(),
            'description':         self.read_string(),

            'tempo':               self.read_numeric(SHORT) / 100.0,
            'auto_save':           self.read_numeric(BYTE) == 1,
            'auto_save_duration':  self.read_numeric(BYTE),
            'time_signature':      self.read_numeric(BYTE),

            'minutes_spent':       self.read_numeric(INT),
            'left_clicks':         self.read_numeric(INT),
            'right_clicks':        self.read_numeric(INT),
            'blocks_added':        self.read_numeric(INT),
            'blocks_removed':      self.read_numeric(INT),
            'song_origin':         self.read_string(),
            'loop':                self.read_numeric(BYTE) == 1 if version >= 4 else False,
            'max_loop_count':      self.read_numeric(BYTE)      if version >= 4 else 0,
            'loop_start':          self.read_numeric(SHORT)     if version >= 4 else 0
        }

    def parse_notes(self, version):
        for current_tick in self.jump():
            for current_layer in self.jump():
                instrument = self.read_numeric(BYTE)
                key        = self.read_numeric(BYTE)
                velocity   = self.read_numeric(BYTE)       if version >= 4 else 100
                panning    = self.read_numeric(BYTE) - 100 if version >= 4 else 0
                pitch      = self.read_numeric(SSHORT)     if version >= 4 else 0
                yield Note(current_tick, current_layer, instrument,
                           key, velocity, panning, pitch)

    def parse_layers(self, layers_count, version):
        for i in range(layers_count):
            name    = self.read_string()
            lock    = self.read_numeric(BYTE) == 1  if version >= 4 else False
            volume  = self.read_numeric(BYTE)
            panning = self.read_numeric(BYTE) - 100 if version >= 2 else 0
            yield Layer(i, name, lock, volume, panning)

    def parse_instruments(self, version):
        for i in range(self.read_numeric(BYTE)):
            name       = self.read_string()
            sound_file = self.read_string()
            pitch      = self.read_numeric(BYTE)
            press_key  = self.read_numeric(BYTE) == 1
            yield Instrument(i, name, sound_file, pitch, press_key)


class Writer(object):
    def __init__(self, buff):
        self.buffer = buff

    def encode_file(self, nbs_file, version):
        self.write_header(nbs_file, version)
        self.write_notes(nbs_file, version)
        self.write_layers(nbs_file, version)
        self.write_instruments(nbs_file, version)

    def encode_numeric(self, fmt, value):
        self.buffer.write(fmt.pack(value))

    def encode_string(self, value):
        self.encode_numeric(INT, len(value))
        self.buffer.write(value.encode(encoding='cp1252'))

    def write_header(self, nbs_file, version):
        header = nbs_file.header

        if version > 0:
            self.encode_numeric(SHORT, 0)
            self.encode_numeric(BYTE, version)
            self.encode_numeric(BYTE, header.default_instruments)
        else:
            self.encode_numeric(SHORT, header.song_length)
        if version >= 3:
            self.encode_numeric(SHORT, header.song_length)
        self.encode_numeric(SHORT, header.song_layers)
        self.encode_string(header.song_name)
        self.encode_string(header.song_author)
        self.encode_string(header.original_author)
        self.encode_string(header.description)

        self.encode_numeric(SHORT, int(header.tempo * 100))
        self.encode_numeric(BYTE, int(header.auto_save))
        self.encode_numeric(BYTE, header.auto_save_duration)
        self.encode_numeric(BYTE, header.time_signature)

        self.encode_numeric(INT, header.minutes_spent)
        self.encode_numeric(INT, header.left_clicks)
        self.encode_numeric(INT, header.right_clicks)
        self.encode_numeric(INT, header.blocks_added)
        self.encode_numeric(INT, header.blocks_removed)
        self.encode_string(header.song_origin)

        if version >= 4:
            self.encode_numeric(BYTE, int(header.loop))
            self.encode_numeric(BYTE, header.max_loop_count)
            self.encode_numeric(SHORT, header.loop_start)

    def write_notes(self, nbs_file, version):
        current_tick = -1

        for tick, chord in nbs_file:
            self.encode_numeric(SHORT, tick - current_tick)
            current_tick = tick
            current_layer = -1

            for note in chord:
                self.encode_numeric(SHORT, note.layer - current_layer)
                current_layer = note.layer
                self.encode_numeric(BYTE, note.instrument)
                self.encode_numeric(BYTE, note.key)
                if version >= 4:
                    self.encode_numeric(BYTE, note.velocity)
                    self.encode_numeric(BYTE, note.panning + 100)
                    self.encode_numeric(SSHORT, note.pitch)

            self.encode_numeric(SHORT, 0)
        self.encode_numeric(SHORT, 0)

    def write_layers(self, nbs_file, version):
        for layer in nbs_file.layers:
            self.encode_string(layer.name)
            if version >= 4:
                self.encode_numeric(BYTE, int(layer.lock))
            self.encode_numeric(BYTE, layer.volume)
            if version >= 2:
                self.encode_numeric(BYTE, layer.panning + 100)

    def write_instruments(self, nbs_file, version):
        self.encode_numeric(BYTE, len(nbs_file.instruments))
        for instrument in nbs_file.instruments:
            self.encode_string(instrument.name)
            self.encode_string(instrument.file)
            self.encode_numeric(BYTE, instrument.pitch)
            self.encode_numeric(BYTE, int(instrument.press_key))
